sessiontxn/isolation/repeatable_read.go

```diff
@@ -0,0 +1,267 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package isolation
+
+import (
+	"context"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/parser/terror"
+	plannercore "github.com/pingcap/tidb/planner/core"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/sessiontxn"
+	"github.com/pingcap/tidb/util/logutil"
+	tikverr "github.com/tikv/client-go/v2/error"
+	"go.uber.org/zap"
+)
+
+// PessimisticRRTxnContextProvider provides txn context for isolation level repeatable-read
+type PessimisticRRTxnContextProvider struct {
+	baseTxnContextProvider
+
+	// Used for ForUpdateRead statement
+	forUpdateTS uint64
+	// It may decide whether to update forUpdateTs when calling provider's getForUpdateTs
+	// See more details in the comments of optimizeWithPlan
+	followingStmtIsPointGetForUpdate bool
+}
+
+// NewPessimisticRRTxnContextProvider returns a new PessimisticRRTxnContextProvider
+func NewPessimisticRRTxnContextProvider(sctx sessionctx.Context, causalConsistencyOnly bool) *PessimisticRRTxnContextProvider {
```

Comment:
Can we replace some usage from `TxnCtx.GetForUpdateTS` to `TxnManager.GetStmtForUpdateTS` in executorBuilder because all pessimistic provider are refactored after it.

Reply:
It may involve many  modifications and I think it's better to do it in another PR.